cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,698 @@
+      }
+    }
+    return Status::OK();
+  }
+
+  Status append_map_batch(const liborc::Type* type, ArrayBuilder* abuilder,
+                          liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                          int64_t length) {
+    auto list_builder = static_cast<ListBuilder*>(abuilder);
+    auto struct_builder = static_cast<StructBuilder*>(list_builder->value_builder());
+    auto batch = static_cast<liborc::MapVectorBatch*>(cbatch);
+    auto keys = batch->keys.get();
+    auto vals = batch->elements.get();
+    auto keytype = type->getSubtype(0);
+    auto valtype = type->getSubtype(1);
+
+    for (int i = offset; i < length + offset; i++) {
+      RETURN_NOT_OK(list_builder->Append());
+      int64_t start = batch->offsets[i];
+      int64_t list_length = batch->offsets[i + 1] - start;
+      if (list_length && (!batch->hasNulls || batch->notNull[i])) {
+        RETURN_NOT_OK(struct_builder->Append(list_length, nullptr));
+        RETURN_NOT_OK(append_batch(keytype, struct_builder->field_builder(0), keys, start,
+                                   list_length));
+        RETURN_NOT_OK(append_batch(valtype, struct_builder->field_builder(1), vals, start,
+                                   list_length));
+      }
+    }
+    return Status::OK();
+  }
+
+  template <class builder_type, class batch_type, class elem_type>
+  Status append_numeric_batch(ArrayBuilder* abuilder, liborc::ColumnVectorBatch* cbatch,
+                              int64_t offset, int64_t length) {
+    auto builder = static_cast<builder_type*>(abuilder);
+    auto batch = static_cast<batch_type*>(cbatch);
+
+    if (length == 0) {
+      return Status::OK();
+    }
+    const uint8_t* valid_bytes = nullptr;
+    if (batch->hasNulls) {
+      valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+    }
+    const elem_type* source = batch->data.data() + offset;
+    RETURN_NOT_OK(builder->Append(source, length, valid_bytes));
+    return Status::OK();
+  }
+
+  template <class builder_type, class target_type, class batch_type, class source_type>
+  Status append_numeric_batch_cast(ArrayBuilder* abuilder,
+                                   liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                                   int64_t length) {
+    auto builder = static_cast<builder_type*>(abuilder);
+    auto batch = static_cast<batch_type*>(cbatch);
+
+    if (length == 0) {
+      return Status::OK();
+    }
+    int start = builder->length();
+
+    const uint8_t* valid_bytes = nullptr;
+    if (batch->hasNulls) {
+      valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+    }
+    RETURN_NOT_OK(builder->AppendNulls(valid_bytes, length));
+
+    const source_type* source = batch->data.data() + offset;
+    target_type* target = reinterpret_cast<target_type*>(builder->data()->mutable_data());
+
+    for (int i = 0; i < length; i++) {
+      target[start + i] = (target_type)source[i];
```

Comment:
`liborc` upcasts all numerics to `int64`, which means we need to downcast back into the arrow structure. This was the best pattern I found to support this, but it gets a bit into the internal details of the `ArrayBuilder`.

Reply:
could do this as `std::copy(source, source + length, target + start)` as well

Reply:
+1 (and if not using `std::copy`, should use `static_cast`)

Reply:
The pattern we used in parquet-cpp was to copy the values into a temporary buffer (that gets reused), but this is OK, too. However, we had not envisioned users directly writing into the internal builder buffer